rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,212 @@
+        self.auto_display = auto_display
+        self.fig = go.Figure()
+        self.layout_kwargs = layout_kwargs if layout_kwargs is not None else {}
+
+        self._validate_input_data()
+
+    @classmethod
+    def construct(
+        cls,
+        models_metrics: pd.DataFrame,
+        show_legend: bool = True,
+        auto_display: bool = True,
+        layout_kwargs: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ) -> "MetricsApp":
+        r"""
+        Construct interactive widget for metric-to-metric trade-off analysis.
+
+        Parameters
+        ----------
+        models_metrics : pd.DataFrame
+            A pandas DataFrame containing metrics for visualization. Required columns:
+                - `Columns.Models` - model names
+                - `Columns.Split` - fold number
+                - Any other numeric columns which represent metric values
+        show_legend : bool, default True
+            Specifies whether to display the chart legend.
+        auto_display : bool, default True
+            Automatically displays the widgets immediately after initialization.
+         layout_kwargs : tp.Optional[tp.Dict[str, tp.Any]], optional, default None
+            Additional arguments from `plotly.graph_objects.Layout`
+
+        Returns
+        -------
+        MetricsApp
+            An instance of `MetricsApp`, providing interactive Jupyter widget for metric visualization.
+
+        Examples
+        --------
+        Create interactive widget
+
+        >>> example_df = pd.DataFrame(
+        ...    {
+        ...        Columns.Model: ["Model1", "Model2", "Model1", "Model2", "Model1", "Model2"],
+        ...        Columns.Split: [0, 0, 1, 1, 2, 2],
+        ...        "prec@10": [0.031, 0.025, 0.027, 0.21, 0.031, 0.033],
+        ...        "recall@10": [0.041, 0.045, 0.055, 0.08, 0.036, 0.021],
+        ...        "novelty@10": [2.6, 11.3, 4.3, 9.8, 3.3, 11.2],
+        ...    })
+        >>> app = MetricsApp.construct(
+        ...    models_metrics=example_df,
+        ...    show_legend=True,
+        ...    auto_display=False,
+        ...    layout_kwargs={"width": 800, "height": 600})
+
+        Get plotly chart from widget state
+
+        >>> fig = app.fig
+        >>> fig = fig.update_layout(
+        ...    title="Metrics: recall@10 vs novelty@10",
+        ...    xaxis_title="novelty@10",
+        ...    yaxis_title="recall@10",
+        ...    legend_title="Models",
+        ... )
+        """
+        app = cls(models_metrics, show_legend, auto_display, layout_kwargs)
+        if auto_display:
+            app.display()
+        return app
+
+    @property
+    def metric_names(self) -> tp.List[str]:
+        """List of metric column names from the `models_metrics`."""
+        non_metric_columns = {Columns.Split, Columns.Model}
+        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+
+    @property
+    def model_names(self) -> tp.List[str]:
```

Comment:
You're using this property quite often, but they theoretically may be slow and may cause problems during the chart updates. Let's add `lru_cache` here.
(Maybe for prev. and next one as well)

Reply:
Done